[MAINTENANCE] ID/PK result_format documentation update (#6716)
great-expectations/great_expectations: docs/reference/expectations/result_format.md

The `result_format` parameter may be either a string or a dictionary which specifies the fields to return in `result`.

>   * For string usage, see [`result_format` values and fields](#result_format-values-and-fields).

Reviewer: Nit - What about limiting the link to the text after the inline code snippet? It just gets rendered a little wonky (with the underline and the underscore of `result_format`. 
![image](https://user-images.githubusercontent.com/9903066/211339087-d3696bec-96e6-4b58-ae4b-a405b5af800e.png)

```suggestion
  * For string usage, see `result_format` [values and fields](#result_format-values-and-fields).
```

>   * For dictionary usage, `result_format` which may include the following keys:

Reviewer 2: This item doesn't read well -- the word "which" makes `result_format` seem insufficient of a subject.

I would change it to:
```
  * For dictionary usage, `result_format` may include the following keys:
```
or:
```
  * For dictionary usage, provide `result_format` as a dictionary, which may include the following keys:
```

(I think I prefer the first of the two.)

Author: Done. Thanks for the catch :)

    * `result_format`: Sets the fields to return in result.

>     * `unexpected_index_column_names`: Defines the primary key (PK) columns used to represent unexpected results.

Reviewer: How about something like the following? Just to be more general that a column does not have to be the primary key column and explicit that multiple columns can be used.
```suggestion
    * `unexpected_index_column_names`: Defines columns that can be used to identify unexpected results, for example primary key (PK) column(s) or other columns with unique identifiers. Supports multiple column names as a list.
```

Author: Love this change. thank you :)

>     * `return_unexpected_index_query`: When running validations, a query (or a set of indices) will be returned that will
>       allow you to retrieve the full set of unexpected results defined by the PK.  Setting this value to `False` will 
>       suppress the output (default is `True`).

Reviewer: Nit - What do you think of this?
```suggestion
    * `return_unexpected_index_query`: When running validations, a query (or a set of indices) will be returned that will
      allow you to retrieve the full set of unexpected results including any columns identified in `unexpected_index_column_names`.  Setting this value to `False` will 
      suppress the output (default is `True`).
```

Author: 👍🏼

    * `partial_unexpected_count`: Sets the number of results to include in partial_unexpected_count, if applicable. If 
      set to 0, this will suppress the unexpected counts.
    * `include_unexpected_rows`: When running validations, this will return the entire row for each unexpected value in
      dictionary form. When using `include_unexpected_rows`, you must explicitly specify `result_format` as well, and
      `result_format` must be more verbose than `BOOLEAN_ONLY`. *WARNING: *
  
  :::warning
  `include_unexpected_rows` returns EVERY row for each unexpected value; for large tables, this could return an 
  unwieldy amount of data.
  :::

## Configure Result Format
Result Format can be applied to either a single Expectation or an entire Checkpoint. When configured at the Expectation-level, 
the configuration will not be persisted, and you will receive a `UserWarning`. We therefore recommend that the Expectation-level
configuration be used for exploratory analysis, with the final configuration added at the Checkpoint-level.

### Expectation Level Config

> To apply `result_format` to an Expectation, pass it into the Expectation. We will first need to obtain a Validator object instance by running the `$ great_expectations suite new` command.

Reviewer: Nit: Just set off the CLI in a parenthetical since there are other ways to set up a validator (the CLI just generating the boilerplate).
```suggestion
To apply `result_format` to an Expectation, pass it into the Expectation. We will first need to obtain a Validator object instance (e.g. by running the `$ great_expectations suite new` command).
```

```python name="result_format_complete_example_set"
```

### Checkpoint Level Config

> To apply `result_format` to every Expectation in a Suite, define it in your Checkpoint configuration under the `runtime_configuration` key.

Reviewer: Just a related question: Does this also work if there are multiple validations (e.g. each with an expectation suite)? Can you specify a validation-specific result_format?

Author: Yes they "technically" work, but will not be persisted in the same way that a checkpoint-level configuration will. Basically everything gets munged together at the graph-validate step when the metrics are collected 😓

```python name="result_format_checkpoint_example"
```

The results will then be stored in the Validation Result after running the Checkpoint.
:::note

> The `unexpected_index_list`, as represented by primary key (PK) columns, is 

Reviewer: ```suggestion
The `unexpected_index_list`, as represented by columns selected via `unexpected_index_column_names`, is 
```

|    SUMMARY                            | Detailed exploratory work with follow-on investigation.      |
|    COMPLETE                           | Debugging pipelines or developing detailed regression tests. |

## Examples

The following examples will use the data defined in the following Pandas DataFrame:

> ```python name="pandas_df_for_result_format"

Reviewer: Totally optional nitpick - do you think this dataframe can be reduced in size so that each of the columns can fit on a single line e.g. `"my_var": ["A", "B", "B", "C", "C", "D", ... <etc>],` instead of 
```
"my_var": [
            "A",
            "B",
            "B",
            "C",
            "C",
            "C",
            "D",
            "D",
            "D",
            "D",
            "E",
            "E",
            "E",
            "E",
            "E",
        ],
```

Author: Done :) This was sortof bugging me too

```

### Behavior for `BOOLEAN_ONLY`

When the `result_format` is `BOOLEAN_ONLY`, no `result` is returned. The result of evaluating the Expectation is  
exclusively returned via the value of the `success` parameter.

For example:

```python name="result_format_boolean_example"
```

Will return the following output:
```python
>>> print(validation_result.success)
False

>>> print(validation_result.result)

> {}

Reviewer: Same comment about adding these to the test script (see below).

**Note:** When unexpected values are duplicated, `unexpected_list` will contain multiple copies of the value.

For example:

```python name="result_format_basic_example_set"
```

Will return the following output:
```python
>>> print(validation_result.success)
False

>>> print(validation_result.result)
{
    "element_count": 15,
    "unexpected_count": 6,
    "unexpected_percent": 40.0,
    "partial_unexpected_list": ["A", "E", "E", "E", "E", "E"],
    "missing_count": 0,
    "missing_percent": 0.0,
    "unexpected_percent_total": 40.0,
    "unexpected_percent_nonmissing": 40.0,
}

> ```

Reviewer: Same comment about adding these to the test script (see below).

For example:

```python name="result_format_basic_example_agg"
```

Will return the following output:
```python
>>> print(validation_result.success)
True

>>> print(validation_result.result)
{'observed_value': 3.6666666666666665}

> ```

Reviewer: Same comment about adding these to the test script (see below).

        "partial_unexpected_list" : [A list of up to 20 values that violate the expectation] (also in `BASIC`)
        'missing_count': The number of missing values in the column
        'missing_percent': The total percent of missing values in the column
        'partial_unexpected_counts': [{A list of objects with value and counts, showing the number of times each of the unexpected values occurs}

>         'partial_unexpected_index_list': [A list of up to 20 of the indices of the unexpected values in the column, as defined by the primary key column]

Reviewer: ```suggestion
        'partial_unexpected_index_list': [A list of up to 20 of the indices of the unexpected values in the column, as defined by the columns in `unexpected_index_column_names`]
```

    }
}
```

For example:
```python name="result_format_summary_example_set"
```

Will return the following output:
```python
>>> print(validation_result.success)
False

>>> print(validation_result.result)
{

>     "element_count": 15,

Reviewer: Should this example (and others for validation_result.success and .result) come from snippets too? That way they are under test and then change automatically if the test script changes? Maybe an assert statement in the script below the script tag if we don't want to show the assert in the docs? E.g. 
```
<snippet name="some name">
expected_validation_result = {
   ... (fill this in)
}
</snippet>
assert validation_result.result == expected_validation_result
```

Author: ![Screen Shot 2023-01-09 at 11 18 52 PM](https://user-images.githubusercontent.com/25670697/211486492-c1985bf3-410d-4e9e-9a6a-789c08c5a9bd.png)

I had a stab at adjusting the output, and there is potentially a small inconsistency (in the formatting) that this introduces. It's pretty minor, so let me know if which direction you would like to go.  I'm happy either way.

Reviewer: How do we do this in other docs? I don't recall the repl style `>>>` in other docs. Maybe we can just show the assert statement? like:
```
assert validation_result.success
assert validation_result.result == {"observed_value": 3.6666666666666665}
```
I see we do that in a few places e.g. https://docs.greatexpectations.io/docs/guides/connecting_to_your_data/advanced/how_to_configure_a_dataconnector_for_splitting_and_sampling_tables_in_sql/#splitting-and-sampling

Author: Great idea. I think this was the result of tension between preserving the existing docs as much as possible (which used the `>>>`) and making them accurate. I like the idea of displaying the `assert`. Update coming up ;)

Author: ![Screen Shot 2023-01-10 at 12 24 15 PM](https://user-images.githubusercontent.com/25670697/211654493-062c3219-a75e-4bfa-b29e-fc579b3f08d3.png)
Boom

    "unexpected_percent_total": 40.0,
    "unexpected_percent_nonmissing": 40.0,
    "partial_unexpected_index_list": [
        {"my_var": "A", "pk_column": "zero"},
        {"my_var": "E", "pk_column": "ten"},
        {"my_var": "E", "pk_column": "eleven"},
        {"my_var": "E", "pk_column": "twelve"},
        {"my_var": "E", "pk_column": "thirteen"},
        {"my_var": "E", "pk_column": "fourteen"},
    ],
    "partial_unexpected_counts": [
        {"value": "E", "count": 5},
        {"value": "A", "count": 1},
    ],
}
```

`column_aggregate_expectation` computes a single aggregate value for the column, and so returns a `observed_value` 
to justify the expectation result. It also includes additional information regarding observed values and counts, 
depending on the specific expectation.

The summary `result` includes:

```python
{
    'success': False,
    'result': {
        'observed_value': The aggregate statistic computed for the column (also in `BASIC`)
        'element_count': The total number of values in the column
        'missing_count':  The number of missing values in the column
        'missing_percent': The total percent of missing values in the column

>         'details': {<expectation-specific result justification fields>}

Reviewer: Oh wow - good catch.

> {'observed_value': 3.6666666666666665}

Reviewer: Similar comment about adding these to the test script.

    'success': False,
    'result': {
        "unexpected_list" : [A list of all values that violate the expectation]

>         'unexpected_index_list': [A list of the indices of the unexpected values in the column, as defined by primary key column]

Reviewer: ```suggestion
        'unexpected_index_list': [A list of the indices of the unexpected values in the column, as defined by the columns in `unexpected_index_column_names`]
```

```

For example:
```python name="result_format_complete_example_set"
```

Will return the following output:
```python
>>> print(validation_result.success)
False

>>> print(validation_result.result)

> {

Reviewer: Same comment about adding these to the test script.

    'success': False,
    'result': {
        'observed_value': The aggregate statistic computed for the column (also in `SUMMARY`)
        'element_count': The total number of values in the column (also in `SUMMARY`)
        'missing_count':  The number of missing values in the column (also in `SUMMARY`)

>         'missing_percent': The total percent of missing values in the column (also in `SUMMARY`)

Reviewer: Good catch. Just to verify - these don't appear for all expectations? Any idea why they were included here (i.e. were they available at one point but not any longer and we didn't update the docs)?

```

For example:
```python name="result_format_complete_example_agg"
```

Will return the following output:
```python
>>> print(validation_result.success)
True

>>> print(validation_result.result)
{'observed_value': 3.6666666666666665}

> ```

Reviewer: Same comment about adding these to the test script.
